File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    answer = 0
    loca = [0, 0]
    visited = []
    for i in dirs:
        vist = False
        temp_loca = loca[:]
        
        if i == "R" and loca[0] < 5:
            loca[0] += 1
        elif i == "L" and loca[0] > -5:
            loca[0] -= 1
        elif i == "U" and loca[1] < 5:
            loca[1] += 1
        elif i == "D" and loca[1] > -5:
            loca[1] -= 1
        else:
            continue
            
        move = (tuple(temp_loca), tuple(loca))
        reverse_move = (tuple(loca), tuple(temp_loca))
        
        if move not in visited and reverse_move not in visited:
            answer += 1
            visited.append(move)
            visited.append(reverse_move)
        
    return answer

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int solution(String dirs) {
        int answer = 0;

        String[] dirsArr = dirs.split("");
		Set<String> set = new HashSet<String>();
		
		int x = 0;
		int y = 0;
		
		for (int i = 0; i < dirsArr.length; i++) {
            int tempX = x;
            int tempY = y;
			switch (dirsArr[i]) {
			case "U":
				if(y+1 <= 5) y++;
				break;
			case "D":
				if(y-1 >= -5) y--;
				break;
			case "L":
				if(x-1 >= -5) x--;
				break;
			case "R":
				if(x+1 <= 5) x++;
				break;
			}

            String key = "(" + tempX + "," + tempY + ")(" + x + "," + y + ")";       
            String reverseKey = "(" + x + "," + y + ")(" + tempX + "," + tempY + ")";
            if (!key.equals(reverseKey)) { 
                set.add(key);
                set.add(reverseKey);
            }
        }
		
		answer = set.size()/2;
        return answer;
	}
}